File: SEMESTER3/PemrogramanBerorientasiObjek/Week8/TUGAS/210059_UTSPBO_Soal4.py
# Program : Soal 3
# Nama    : Prames Ray Lapian
# NPM     : 140810210059
# Kelas   : A
# Tanggal : 23 Oktober 2022

class Waktu:
    __h = 0
    __m = 0
    __s = 0
    
    def __init__(self, h, m, s):
        self.__h = int(h)
        self.__m = int(m)
        self.__s = int(s)
        
    def setJam(self, h):
        self.__h = int(h)
    
    def setMenit(self, m):
        self.__m = int(m)

    def setDetik(self, s):
        self.__s = int(s)

    def getJam(self):
        return self.__h

    def getMenit(self):
        return self.__m

    def getDetik(self):
        return self.__s
    
    def inputWaktu(self):
        self.__h = int(input("Masukkan jam\t: "))
        self.__m = int(input("Masukkan menit\t: "))
        self.__s = int(input("Masukkan detik\t: "))

    def getWaktu(self):
        jam = ""
        menit = ""
        detik = ""
        if(self.__h<10):
            jam = "0"
        if(self.__m<10):
            menit = "0"
        if(self.__s<10):
            detik = "0"
        return jam + str(self.__h) + ":" + menit + str(self.__m) + ":" + detik + str(self.__s)

    def detikTotal(self):
        total = int(self.__h*3600 + self.__m*60 + self.__s)
        return total
    
    def durasi(self, keluar):
        durasi = Waktu(0,0,0)
        detikMasuk = int(self.detikTotal())
        detikKeluar = int(keluar.detikTotal())
        total =  detikKeluar - detikMasuk
        durasi.konversi(total)
        return durasi

    def konversi(self, detik):
        self.__h = int(detik/3600)
        detik = detik%3600
        self.__m = int(detik/60)
        detik = detik%60
        self.__s = int(detik)

class Kendaraan:
    __plat = ""
    __jenis = ""
    __masuk = Waktu(0,0,0)
    __keluar = Waktu(0,0,0)

    def __init__(self, plat, jenis):
        self.__plat = str(plat)
        self.__jenis = str(jenis)

    def setPlat(self, plat):
        self.__plat = str(plat)

    def setJenis(self, jenis):
        self.__jenis = str(jenis)

    def getPlat(self):
        return self.__plat

    def getJenis(self):
        return self.__jenis

    def getMasuk(self):
        return self.__masuk
    
    def getKeluar(self):
        return self.__keluar
    
    def inputKendaraan(self):
        self.__plat = str(input("Masukkan Plat Kendaraan : "))
        self.__jenis = str(input("Masukkan Jenis Kendaraan : "))
        print("Masuk")
        self.__masuk.inputWaktu()
        print("Keluar")
        self.__keluar.inputWaktu()

    def durasiParkir(self):
        temp = Waktu(0,0,0)
        temp = self.__masuk.durasi(self.__keluar)
        if (temp.getMenit() > 10):
                temp.setJam(temp.getJam()+1)
        return temp
    
    def biaya(self):
        biaya = int(0)
        if(self.durasiParkir().detikTotal() > 600):
            match self.__jenis:
                case "Motor":
                    biaya = (2000*self.durasiParkir().getJam())
                case "Mobil":
                    biaya = (3000*self.durasiParkir().getJam())
        return biaya

class Larik:
    __banyak = int(0)
    __kendaraan = []

    def __init__(self, banyak):
        self.__banyak = int(banyak)

    def setBanyak(self, banyak):
        self.__banyak = int(banyak)

    def getBanyak(self):
        return self.__banyak

    def inputLarik(self):
        i = 0
        while(i<self.__banyak):
            print("Kendaraan ke -" , (i+1))
            x = Kendaraan(" ", " ")
            x.inputKendaraan()
            self.__kendaraan.append(x)
            i = i+1

    def printLarik(self):
        print(" [List Parkir Kendaraan] ")
        no = 1
        if(self.__kendaraan[0] == " "):
            print("NULL")
        else:
            i = 0
            print("-------------------------------------------------------------------------------------------------------------------")
            print("No\tNo Kendaraan\tJenis\t\t Masuk\t\t Keluar\t\t Durasi\t\tJam Terhitung\tBiaya\t")
            print("-------------------------------------------------------------------------------------------------------------------")
            while(i<self.__banyak):
                if(self.__kendaraan[0].getPlat() == " "):
                    break
                else:
                    print(
                        no , "\t" ,
                        self.__kendaraan[i].getPlat(), "\t", 
                        self.__kendaraan[i].getJenis(), "\t\t", 
                        self.__kendaraan[i].getMasuk().getWaktu(), "\t", 
                        self.__kendaraan[i].getKeluar().getWaktu(), "\t", 
                        self.__kendaraan[i].durasiParkir().getWaktu(), "\t", 
                        self.__kendaraan[i].durasiParkir().getJam(), "\t\t", 
                        self.__kendaraan[i].biaya()
                    )
                    no = no +1
                i = i+1
            print("-------------------------------------------------------------------------------------------------------------------")

    def totalBiaya(self):
        hasil = 0
        i = 0
        while (i<self.__banyak):
            hasil = hasil + self.__kendaraan[i].biaya()
            i = i+1
        print("Total Biaya Parkir =", hasil)


# main ------------------------------------------------------------------------------------------
banyak = int(input("Masukkan banyak kendaraan : "))
print()

kendaraan = Larik(banyak)

kendaraan.setBanyak(banyak)
kendaraan.inputLarik()
print()
kendaraan.printLarik()
kendaraan.totalBiaya()
    

File: SEMESTER3/PemrogramanBerorientasiObjek/Week8/TUGAS/210059_UTSPBO_Soal4.cpp
// Program : Soal 3
// Nama    : Prames Ray Lapian
// NPM     : 140810210059
// Kelas   : A
// Tanggal : 23 Oktober 2022

#include <iostream>
using namespace std;

class Waktu{
    private :
        int h, m, s;

    public :
        void setJam(int jam){
            h = jam;
        }

        void setMenit(int mnt){
            m = mnt;
        }

        void setDetik(int dtk){
            s = dtk;
        }
        
        int getJam(){
            return h;
        }
        
        int getMenit(){
            return m;
        }
        
        int getDetik(){
            return s;
        }

        void inputWaktu(string text){
            cout << "Masukkan Jam " << text << "\t: ";
            cin >> h;
            cout << "Masukkan Menit " << text << "\t: ";
            cin >> m;
            cout << "Masukkan Detik " << text << "\t: ";
            cin >> s;
        }

        string getWaktu(){
            string jam, menit, detik;
            jam = "";
            menit = "";
            detik = "";
            if(h<10){
                jam = "0";
            } else{}
            if(m<10){
                menit = "0";
            } else{}
            if(s<10){
                detik = "0";
            } else{}
            return (jam + to_string(h) + ":" + menit + to_string(m) + ":" + detik + to_string(s));
        }

        int detikTotal(){
            int total = (h*3600 + m*60 + s);
            return total;
        }
        
        Waktu durasi(Waktu keluar){
            Waktu durasi;
            int detikMasuk = this->detikTotal();
            int detikKeluar = keluar.detikTotal();
            int total = detikKeluar - detikMasuk;
            durasi.konversi(total);
            return durasi;
        }

        void konversi(int detik){
            h = detik/3600;
            detik = detik%3600;
            m = detik/60;
            detik = detik%60;
            s = detik;
        }
};

class Kendaraan{
    private :
        string plat;
        int jenis;
        Waktu masuk;
        Waktu keluar;
        
    public :
        void setPlat(string platK){
            plat = platK;
        }

        void setJenis(int jenisK){
            jenis = jenisK;
        }

        string getPlat(){
            return plat;
        }

        int getJenis(){
            return jenis;
        }

        Waktu getMasuk(){
            return masuk;
        }
        
        Waktu getKeluar(){
            return keluar;
        }
        
        void inputKendaraan(){
            cout << "Masukkan No Kendaraan : ";
            cin >> plat;
            cout << "Masukkan Jenis Kendaraan" << endl
                 << "1. Mobil" << endl  
                 << "2. Motor" << endl
                 << "Opsi: ";
            cin >> jenis;
            masuk.inputWaktu("Masuk");
            keluar.inputWaktu("Keluar");
        }

        int tarifParkir(){
            int tarif = 0;
            if (jenis == 2){
                tarif = 2000;
            }
            else if (jenis == 1){
                tarif = 3000;
            }
            return tarif;
        }
        
        Waktu durasiParkir(){
            Waktu durpak;
            durpak = masuk.durasi(keluar);
            if (durpak.getMenit() > 10){
                durpak.setJam(durpak.getJam()+1);
            }
            return durpak;
        }
        
        int getBiaya(){
            int biaya;
            if(durasiParkir().detikTotal() > 600){
                if (jenis == 2){
                    biaya = (2000*durasiParkir().getJam());
                }
                else if (jenis == 1){
                    biaya = (3000*durasiParkir().getJam());
                }
            }
            return biaya;
        }
};

class Larik{
    private:
        int banyak;
        Kendaraan kendaraan[10];
    
    public:
        void setBanyak(int ukuran){
            banyak = ukuran;
        }

        void inputLarik(){
            for(int i=0; i<banyak; i++){
                cout << endl << "Kendaraan ke-" << i+1 << "\n";
                kendaraan[i].inputKendaraan();
            }
        }

        int getBanyak(){
            return banyak;
        }

        void print(){
            int no = 1;
            cout<<"\t\t\t\t\t[List Parkir Kendaraan] \n";
            if(kendaraan[0].getPlat() == " "){
                cout<<"NULL\n";
            }
            else{
                cout<<"----------------------------------------------------------------------------------------------------------------------\n";
                cout<<"No\tNo Kendaraan\t\tJenis\t\t Masuk\t\t Keluar\t\t Durasi\t\tLama Jam\tBiaya\t\n";
                cout<<"----------------------------------------------------------------------------------------------------------------------\n";
                for(int i=0; i<banyak; i++){
                    if(kendaraan[i].getPlat() == " "){
                        break;
                    }
                    else{
                        cout<<
                            no << "\t" <<
                            //identitas
                            kendaraan[i].getPlat() << "\t\t" << kendaraan[i].getJenis() << "\t\t" << 
                            //waktu masuk keluar
                            kendaraan[i].getMasuk().getWaktu() << "\t" << kendaraan[i].getKeluar().getWaktu() << "\t" <<
                            //waktu parkir 
                            kendaraan[i].durasiParkir().getWaktu() << "\t" << kendaraan[i].durasiParkir().getJam() << "\t\t " << 
                            //tarif parkir
                            kendaraan[i].getBiaya() << "\t\n";
                        no++;
                    }
                }
                cout<<"----------------------------------------------------------------------------------------------------------------------\n";
            }
        }
        
        void totalBiaya(){
            int hasil = 0;
            for(int i=0; i<banyak; i++){
                hasil += kendaraan[i].getBiaya();
            }
            cout << "Total Biaya Parkir = " << hasil;
        }
};

main(){
    int banyak;
    cout << "Masukkan banyak kendaraan : ";
    cin >> banyak;
    cout << endl;

    Larik kendaraan;
    kendaraan.setBanyak(banyak);
    kendaraan.inputLarik();
    cout << endl;
    kendaraan.print();
    kendaraan.totalBiaya();
}